Dart/src/com/jetbrains/lang/dart/ide/formatter/DartSpacingProcessor.java

```diff
@@ -371,14 +349,41 @@ else if (type1 == COLON) {
     }
 
     if (elementType == VALUE_EXPRESSION && type2 == CASCADE_REFERENCE_EXPRESSION) {
-      //if (type1 == CASCADE_REFERENCE_EXPRESSION) {
-      //  if (((AbstractBlock)child1).getNode().getLastChildNode().getElementType() == CALL_EXPRESSION) {
-      //    // TODO if child1 message name == child2 message name, do not add line break
-      //  }
-      //}
+      if (type1 == CASCADE_REFERENCE_EXPRESSION) {
+        if (cascadesAreSameMethod(((AbstractBlock)child1).getNode(), ((AbstractBlock)child2).getNode())) {
+          return Spacing.createSpacing(0, 1, 0, mySettings.KEEP_LINE_BREAKS, mySettings.KEEP_BLANK_LINES_IN_CODE);
+        }
+      }
+      else if (type1 == REFERENCE_EXPRESSION) {
+        CompositeElement elem = (CompositeElement)myNode;
+        ASTNode[] childs = elem.getChildren(CASCADE_REFERENCE_EXPRESSION_SET);
+        if (childs.length == 1 || allCascadesAreSameMethod(childs)) {
+          return Spacing.createSpacing(0, 1, 0, mySettings.KEEP_LINE_BREAKS, mySettings.KEEP_BLANK_LINES_IN_CODE);
+        }
+      }
       return addLineBreak();
     }
 
+    if (type1 == CLOSING_QUOTE && type2 == OPEN_QUOTE && elementType == STRING_LITERAL_EXPRESSION) {
+      ASTNode sib = node1;
+      int preserveNewline = 0;
+      // Adjacent strings on the same line should not be split.
+      while ((sib = sib.getTreeNext()) != null) {
+        // Comments are handled elsewhere.
+        if (sib.getElementType() == WHITE_SPACE) {
+          String ws = sib.getText();
+          if (ws.contains("\n") || ws.contains("\r")) {
+            preserveNewline++;
+            break;
+          }
+          continue;
+        }
+        break;
+      }
+      // Adjacent strings on separate lines should not include blank lines.
+      return Spacing.createSpacing(0, 1, preserveNewline, true, 0);
```

Comment:
Shouldn't the first argument be 1?
I guess there are many more places in this class where the first argument of Spacing.createSpacing() should be equal to the second one, but 0 and 1 are passed instead.